### ISSUE TITLE
ngx_pagespeed fails to build with nginx 1.7.11 and --with-threads

### ISSUE THREAD

**Issue by crowell** [pagespeed/ngx_pagespeed#948]
bug report from here : https://groups.google.com/forum/#!topic/ngx-pagespeed-discuss/62fSCt3av9A
configure nginx --with-threads
eg

``
CC=clang CXX=clang++ ./configure --with-threads --add-module=/tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta
``

```
jcrowell0 gnx/nginx-1.7.11 » make
make -f objs/Makefile
make[1]: Entering directory `/tmp/gnx/nginx-1.7.11'
clang -c -pipe  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Werror -g    -I src/core -I src/event -I src/event/modules -I src/os/unix -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/chromium/src -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/google-sparsehash/src -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/protobuf/src -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/re2/src -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/out/Release/obj/gen -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/out/Release/obj/gen/protoc_out/instaweb -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/apr/src/include -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/aprutil/src/include -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/apr/gen/arch/linux/x64/include -I /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/psol/include/third_party/aprutil/gen/arch/linux/x64/include -I objs -I src/http -I src/http/modules -I src/mail \
                -o objs/addon/src/ngx_message_handler.o \
                /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.h:21:
In file included from src/core/ngx_core.h:49:
src/os/unix/ngx_files.h:387:25: error: unknown type name 'ngx_thread_task_t'; did you mean 'ngx_thread_cond_t'?
ssize_t ngx_thread_read(ngx_thread_task_t **taskp, ngx_file_t *file,
                        ^~~~~~~~~~~~~~~~~
                        ngx_thread_cond_t
src/os/unix/ngx_thread.h:28:25: note: 'ngx_thread_cond_t' declared here
typedef pthread_cond_t  ngx_thread_cond_t;
                        ^
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.h:21:
In file included from src/core/ngx_core.h:57:
src/core/ngx_buf.h:105:50: error: unknown type name 'ngx_thread_task_t'; did you mean 'ngx_thread_cond_t'?
    ngx_int_t                  (*thread_handler)(ngx_thread_task_t *task,
                                                 ^~~~~~~~~~~~~~~~~
                                                 ngx_thread_cond_t
src/os/unix/ngx_thread.h:28:25: note: 'ngx_thread_cond_t' declared here
typedef pthread_cond_t  ngx_thread_cond_t;
                        ^
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.h:21:
In file included from src/core/ngx_core.h:57:
src/core/ngx_buf.h:107:5: error: unknown type name 'ngx_thread_task_t'; did you mean 'ngx_thread_cond_t'?
    ngx_thread_task_t           *thread_task;
    ^~~~~~~~~~~~~~~~~
    ngx_thread_cond_t
src/os/unix/ngx_thread.h:28:25: note: 'ngx_thread_cond_t' declared here
typedef pthread_cond_t  ngx_thread_cond_t;
                        ^
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.h:21:
In file included from src/core/ngx_core.h:62:
src/core/ngx_file.h:27:48: error: unknown type name 'ngx_thread_task_t'; did you mean 'ngx_thread_cond_t'?
    ngx_int_t                (*thread_handler)(ngx_thread_task_t *task,
                                               ^~~~~~~~~~~~~~~~~
                                               ngx_thread_cond_t
src/os/unix/ngx_thread.h:28:25: note: 'ngx_thread_cond_t' declared here
typedef pthread_cond_t  ngx_thread_cond_t;
                        ^
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed-release-1.9.32.3-beta/src/ngx_message_handler.h:21:
In file included from src/core/ngx_core.h:83:
src/core/ngx_connection.h:188:5: error: unknown type name 'ngx_thread_task_t'; did you mean 'ngx_thread_cond_t'?
    ngx_thread_task_t  *sendfile_task;
    ^~~~~~~~~~~~~~~~~
    ngx_thread_cond_t
src/os/unix/ngx_thread.h:28:25: note: 'ngx_thread_cond_t' declared here
typedef pthread_cond_t  ngx_thread_cond_t;
                        ^
5 errors generated.
make[1]: *** [objs/addon/src/ngx_message_handler.o] Error 1
make[1]: Leaving directory `/tmp/gnx/nginx-1.7.11'
make: *** [build] Error 2
```

**Comment by crowell** [comment]
applying this diff

```diff
diff --git a/src/log_message_handler.h b/src/log_message_handler.h
index 05f7277..16125de 100644
--- a/src/log_message_handler.h
+++ b/src/log_message_handler.h
@@ -22,6 +22,7 @@
 extern "C" {
   #include <ngx_config.h>
   #include <ngx_core.h>
+  #include <ngx_thred.h>
   #include <ngx_http.h>
   #include <ngx_log.h>
 }
diff --git a/src/ngx_rewrite_driver_factory.h b/src/ngx_rewrite_driver_factory.h
index 9ce36dd..5d97e95 100644
--- a/src/ngx_rewrite_driver_factory.h
+++ b/src/ngx_rewrite_driver_factory.h
@@ -20,6 +20,7 @@
 #define NGX_REWRITE_DRIVER_FACTORY_H_
 
 extern "C" {
+  #include <ngx_thread.h>
   #include <ngx_core.h>
   #include <ngx_http.h>
   #include <ngx_config.h>
```

fixes the build for 1.7.11, but breaks 1.6.2 (the current stable)

**Comment by crowell** [comment]
error from building on 1.6.2 with that patch applied

```
clang -c -pipe  -O -Wall -Wextra -Wpointer-arith -Wconditional-uninitialized -Wno-unused-parameter -Werror -g    -I src/core -I src/event -I src/event/modules -I src/os/unix -I /tmp/gnx/ngx_pagespeed/psol/include -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/chromium/src -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/google-sparsehash/src -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/protobuf/src -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/re2/src -I /tmp/gnx/ngx_pagespeed/psol/include/out/Release/obj/gen -I /tmp/gnx/ngx_pagespeed/psol/include/out/Release/obj/gen/protoc_out/instaweb -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/apr/src/include -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/aprutil/src/include -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/apr/gen/arch/linux/x64/include -I /tmp/gnx/ngx_pagespeed/psol/include/third_party/aprutil/gen/arch/linux/x64/include -I objs -I src/http -I src/http/modules -I src/mail \
                -o objs/addon/src/ngx_message_handler.o \
                /tmp/gnx/ngx_pagespeed/src/ngx_message_handler.cc
In file included from /tmp/gnx/ngx_pagespeed/src/ngx_message_handler.cc:17:
In file included from /tmp/gnx/ngx_pagespeed/src/ngx_message_handler.h:21:
In file included from src/os/unix/ngx_thread.h:13:
In file included from src/core/ngx_core.h:41:
src/core/ngx_rbtree.h:51:24: error: unknown type name 'ngx_thread_volatile'
void ngx_rbtree_insert(ngx_thread_volatile ngx_rbtree_t *tree,
                       ^
src/core/ngx_rbtree.h:51:57: error: expected ')'
void ngx_rbtree_insert(ngx_thread_volatile ngx_rbtree_t *tree,
                                                        ^
src/core/ngx_rbtree.h:51:23: note: to match this '('
void ngx_rbtree_insert(ngx_thread_volatile ngx_rbtree_t *tree,
                      ^
src/core/ngx_rbtree.h:53:24: error: unknown type name 'ngx_thread_volatile'
void ngx_rbtree_delete(ngx_thread_volatile ngx_rbtree_t *tree,
                       ^
src/core/ngx_rbtree.h:53:57: error: expected ')'
void ngx_rbtree_delete(ngx_thread_volatile ngx_rbtree_t *tree,
                                                        ^
src/core/ngx_rbtree.h:53:23: note: to match this '('
void ngx_rbtree_delete(ngx_thread_volatile ngx_rbtree_t *tree,
                      ^
4 errors generated.
make[1]: *** [objs/addon/src/ngx_message_handler.o] Error 1
make[1]: Leaving directory `/tmp/gnx/nginx-1.6.2'
make: *** [build] Error 2
```

**Comment by crowell** [comment]
fixed in a81cc997a96844c9ab77c4c2bb805df455033e28
